homeassistant/config_entries.py

```diff
@@ -0,0 +1,516 @@
+        """
+        component = getattr(hass.components, self.domain)
+
+        supports_unload = hasattr(component, 'async_unload_entry')
+
+        if not supports_unload:
+            return False
+
+        try:
+            result = yield from component.async_unload_entry(hass, self)
+
+            if not isinstance(result, bool):
+                _LOGGER.error('%s.async_unload_entry did not return boolean',
+                              self.domain)
+                result = False
+
+            return result
+        except Exception:  # pylint: disable=broad-except
+            _LOGGER.exception('Error unloading entry %s for %s',
+                              self.title, self.domain)
+            self.state = ENTRY_STATE_FAILED_UNLOAD
+            return False
+
+    def as_dict(self):
+        """Return dictionary version of this entry."""
+        return {
+            'entry_id': self.entry_id,
+            'version': self.version,
+            'domain': self.domain,
+            'title': self.title,
+            'data': self.data,
+            'source': self.source,
+        }
+
+
+class ConfigError(HomeAssistantError):
+    """Error while configuring an account."""
+
+
+class UnknownEntry(ConfigError):
+    """Unknown entry specified."""
+
+
+class UnknownHandler(ConfigError):
+    """Unknown handler specified."""
+
+
+class UnknownFlow(ConfigError):
+    """Uknown flow specified."""
+
+
+class UnknownStep(ConfigError):
+    """Unknown step specified."""
+
+
+class ConfigEntries:
+    """Manage the configuration entries.
+
+    An instance of this object is available via `hass.config_entries`.
+    """
+
+    def __init__(self, hass, hass_config):
+        """Initialize the entry manager."""
+        self.hass = hass
+        self.flow = FlowManager(hass, hass_config, self._async_add_entry)
+        self._hass_config = hass_config
+        self._entries = None
+        self._sched_save = None
+
+    @callback
+    def async_domains(self):
+        """Return domains for which we have entries."""
+        seen = set()
+        result = []
+
+        for entry in self._entries:
+            if entry.domain not in seen:
+                seen.add(entry.domain)
+                result.append(entry.domain)
+
+        return result
+
+    @callback
+    def async_entries(self, domain=None):
+        """Return all entries or entries for a specific domain."""
+        if domain is None:
+            return list(self._entries)
+        return [entry for entry in self._entries if entry.domain == domain]
+
+    @asyncio.coroutine
+    def async_remove(self, entry_id):
+        """Remove an entry."""
+        found = None
+        for index, entry in enumerate(self._entries):
+            if entry.entry_id == entry_id:
+                found = index
+                break
+
+        if found is None:
+            raise UnknownEntry
+
+        entry = self._entries[found]
+        self._entries.pop(found)
```

Comment:
*Suggestion:* `pop()` also returns the removed element, so this could be merged with above statement.

Reply:
`entry = self._entries.pop(found)`